theon/src/allocator.rs

```diff
@@ -94,15 +481,28 @@ mod global {
         }
     }
 
-    unsafe impl<T: Heap> GlobalAlloc for BumpAlloc<T> {
+    /// GlobalQuickAlloc is a wrapper around a QuickFit over a
+    /// GlobalHeap that uses interior mutability to implement
+    /// the GlobalAlloc trait.
+    struct GlobalQuickAlloc(UnsafeCell<QuickFit<GlobalHeap>>);
+
+    unsafe impl GlobalAlloc for GlobalQuickAlloc {
         unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-            self.alloc_bytes(layout.align(), layout.size())
-                .map_or(ptr::null_mut(), |p| p.as_mut_ptr())
+            let quick = unsafe { &mut *self.0.get() };
+            quick.malloc(layout)
+        }
+        unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+            let quick = unsafe { &mut *self.0.get() };
+            quick.free(ptr, layout);
+        }
+        unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+            let quick = unsafe { &mut *self.0.get() };
+            quick.realloc(ptr, layout, new_size)
         }
-        unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
     }
 
     #[global_allocator]
-    static mut BUMP_ALLOCATOR: BumpAlloc<GlobalHeap> =
-        BumpAlloc::new(GlobalHeap([0u8; GLOBAL_HEAP_SIZE]));
+    static mut GLOBAL_ALLOCATOR: GlobalQuickAlloc = GlobalQuickAlloc(UnsafeCell::new(
```

Comment:
So I should have mentioned this at the start, but this is actually designed for a single-threaded context. The only time this allocator will ever be used in a global context is early in boot, when `theon` is running, and by definition `theon` only runs a single thread on a single core.

Regardless, you're right: it should have some kind of protection against error here, and I've implemented that by replacing the UnsafeCell with an `AtomicPtr`. This, however, led to another issue: when I tried to make `GLOBAL_ALLOCATOR` non-mut, Rust gave me an error: "error[E0658]: raw mutable pointers are not allowed in statics".

I'm not sure how best to approach this (but am totally open to ideas).

Reply:
Heh, it wouldn't be the first time I walked into a wall. :-D

I'm fine with `const_mut_refs`, and added it to `main` and removed the `mut` from `GLOBAL_ALLOCATOR`. Curiously, the compiler was happy with it as-is (without having to implement `sync` explicitly; presumably because everything in it is sync?).